util/math_reference.cpp

```diff
@@ -346,7 +356,17 @@ T mul_hi_signed(T a, T b) {
   lo >>= shft;
   hi += lo + (cross1 >> shft) + (cross2 >> shft);
 
-  return (a >> msb) ^ (b >> msb) ? static_cast<T>(~hi) : hi;
+  T result = hi;
+  // if result is negative
+  if ((a >> msb) ^ (b >> msb)) {
```

Comment:
All this obfuscated code does not make sense.
```suggestion
  if ((a < 0) ^ (b < 0)) {
```
Feel free to simplify all the code above too. :-)

Reply:
Done.